src/compiler/checker.ts

```diff
@@ -34158,12 +34158,12 @@ namespace ts {
                     setLastResult(state, /*type*/ undefined);
                     const operator = operatorToken.kind;
                     if (operator === SyntaxKind.AmpersandAmpersandToken || operator === SyntaxKind.BarBarToken || operator === SyntaxKind.QuestionQuestionToken) {
-                        if (operator === SyntaxKind.AmpersandAmpersandToken) {
-                            let parent = node.parent;
-                            while (parent.kind === SyntaxKind.ParenthesizedExpression
-                                || isBinaryExpression(parent) && (parent.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken || parent.operatorToken.kind === SyntaxKind.BarBarToken)) {
-                                parent = parent.parent;
-                            }
+                        let parent = node.parent;
+                        while (parent.kind === SyntaxKind.ParenthesizedExpression
+                            || isBinaryExpression(parent) && (parent.operatorToken.kind === SyntaxKind.AmpersandAmpersandToken || parent.operatorToken.kind === SyntaxKind.BarBarToken)) {
+                            parent = parent.parent;
+                        }
+                        if (operator === SyntaxKind.AmpersandAmpersandToken || isIfStatement(parent)) {

```

Comment:
I've thought that perhaps with some of my recent changes this could be reverted but it seems that it can't. Without this change the error here goes away:
https://github.com/microsoft/TypeScript/blob/8f0a3929b4b60a5aeef9e355036243b92567a214/tests/baselines/reference/truthinessCallExpressionCoercion2.errors.txt#L100-L101

It's quite interesting because a similar error here **doesn't** go away:
https://github.com/microsoft/TypeScript/blob/8f0a3929b4b60a5aeef9e355036243b92567a214/tests/baselines/reference/truthinessCallExpressionCoercion2.errors.txt#L108-L109

So it seems that the logic is very sensitive when it comes to handling left/right side of a binary expression. Should I strive to revert this change here and find another solution? Or is this acceptable?